internal/controllers/gateway/httproute_controller.go

```diff
@@ -217,11 +221,22 @@ func (r *HTTPRouteReconciler) listHTTPRoutesForGatewayClass(ctx context.Context,
 	gateways := make(map[string]map[string]struct{})
 	for _, gateway := range gatewayList.Items {
 		if string(gateway.Spec.GatewayClassName) == gwc.Name {
-			_, ok := gateways[gateway.Namespace]
-			if !ok {
-				gateways[gateway.Namespace] = make(map[string]struct{})
+			// KIC introduced a new flag called `--gateway-namespaced-name`.
+			// If this flag is set, KIC will only reconcile the specified gateway.
+			// https://github.com/Kong/kubernetes-ingress-controller/issues/5322
+			if r.GatewayNN.Namespace != "" && r.GatewayNN.Name != "" {
+				_, ok := gateways[r.GatewayNN.Namespace]
+				if !ok {
+					gateways[r.GatewayNN.Namespace] = make(map[string]struct{})
+				}
+				gateways[r.GatewayNN.Namespace][r.GatewayNN.Name] = struct{}{}
+			} else {
+				_, ok := gateways[gateway.Namespace]
+				if !ok {
+					gateways[gateway.Namespace] = make(map[string]struct{})
+				}
+				gateways[gateway.Namespace][gateway.Name] = struct{}{}
```

Comment:
```suggestion
		// If the GatewayNN is set, only the specified gateway is took into account.
		if onlyGatewayToReconcile, ok := r.GatewayNN.Get(); ok {
			if onlyGatewayToReconcile.Namespace != gateway.Namespace || onlyGatewayToReconcile.Name != gateway.Name {
				continue
			}
		}
		if string(gateway.Spec.GatewayClassName) == gwc.Name {
			_, ok := gateways[gateway.Namespace]
			if !ok {
				gateways[gateway.Namespace] = make(map[string]struct{})
			}
			gateways[gateway.Namespace][gateway.Name] = struct{}{}
```

I think we should ensure that the `gateway` we're iterating over is the desired `r.GatewayNN` and add it to the mapping only if it is.